Ginger/Ginger/UserControlsLib/ActionInputValueUserControlLib/ActionInputValueUserControl.xaml.cs

```diff
@@ -36,17 +40,136 @@ public partial class ActionInputValueUserControl : UserControl
+                }
+                else if(attr.GetType() == typeof(InvalidValueAttribute))
+                {
+                    InvalidValueValidationRule IVR = (InvalidValueValidationRule)xTextBoxInputTextBox.ValueTextBox.GetValidationRule(TextBox.TextProperty, typeof(InvalidValueValidationRule));                   
+                    if(IVR is null)
+                    {
+                        xTextBoxInputTextBox.ValueTextBox.AddValidationRule(new InvalidValueValidationRule(((InvalidValueAttribute)attr).InvalidValue));
+                    }
+                    else
+                    {
+                        IVR.InvalidValue.AddRange(((InvalidValueAttribute)attr).InvalidValue);
+                    }
+                }
+                else if (attr.GetType() == typeof(ValidValueAttribute))
+                {
+                    ValidValueValidationRule VVR = (ValidValueValidationRule)xTextBoxInputTextBox.ValueTextBox.GetValidationRule(TextBox.TextProperty, typeof(ValidValueValidationRule));
+                    if (VVR is null)
+                    {
+                        xTextBoxInputTextBox.ValueTextBox.AddValidationRule(new ValidValueValidationRule(((ValidValueAttribute)attr).ValidValue));
+                    }
+                    else
+                    {
+                        VVR.ValidValues.AddRange(((ValidValueAttribute)attr).ValidValue);
+                    }
+                }                
+            }
+        }
+
+        private void SetParamLayout()
+        {                     
+            // read and process param attrs like Label, Tooltip, Location
+            if (mActionParamProperties == null)
+            {
+                return;
+            }
+            
+            foreach (Attribute attr in mActionParamProperties)
+            {
+                if (attr.GetType() == typeof(LabelAttribute))
+                {
+                    mLabel = Regex.Replace(Regex.Replace(Regex.Replace(((LabelAttribute)attr).Label, @"(\P{Ll})(\P{Ll}\p{Ll})", "$1 $2"), @"(\p{Ll})(\P{Ll})", "$1 $2"), @"\b\w", (Match match) => match.ToString().ToUpper());
+                }
+                else if (attr.GetType() == typeof(TooltipAttribute))
+                {
+                    xTextBoxInputPnl.ToolTip = ((TooltipAttribute)attr).Tooltip;
+                }
+                else if (attr.GetType() == typeof(DefaultAttribute))
+                {
+                    mDefaultValue = ((DefaultAttribute)attr).DefaultValue;
+                }
+                else if (attr.GetType() == typeof(BrowseAttribute))
+                {
+                    mBrowseNeeded = ((BrowseAttribute)attr).IsNeeded;
+                }
+                else if (attr.GetType() == typeof(BrowseTypeAttribute))
+                {
+                    if (((BrowseTypeAttribute)attr).Browseype.ToString() == Ginger.Activities.UCValueExpression.eBrowserType.File.ToString())
```

Comment:
Why we do ToString on enum?

Reply:
in BrowseTypeAttribute i have created eBrowseType enum which is not same as Ginger.Activities.UCValueExpression.eBrowserType